js_modules/dagit/packages/core/src/workspace/WorkspaceContext.tsx

```diff
@@ -273,16 +286,41 @@ const useVisibleRepos = (
     [setHiddenKeys],
   );
 
-  const visibleOptions = React.useMemo(() => {
+  const setVisible = React.useCallback(
+    (repoAddresses: RepoAddress[]) => {
+      const keysToShow = new Set(
+        repoAddresses.map((repoAddress) => `${repoAddress.name}:${repoAddress.location}`),
+      );
+      setHiddenKeys((current) => {
+        return current?.filter((key) => !keysToShow.has(key));
+      });
+    },
+    [setHiddenKeys],
+  );
+
+  const setHidden = React.useCallback(
+    (repoAddresses: RepoAddress[]) => {
+      const keysToHide = new Set(
+        repoAddresses.map((repoAddress) => `${repoAddress.name}:${repoAddress.location}`),
+      );
+      setHiddenKeys((current) => {
+        const updatedSet = new Set([...(current || []), ...keysToHide]);
+        return Array.from(updatedSet);
+      });
+    },
+    [setHiddenKeys],
+  );
+
+  const visibleRepos = React.useMemo(() => {
     // If there's only one repo, skip the local storage check -- we have to show this one.
     if (allRepos.length === 1) {
       return allRepos;
-    } else {
-      return allRepos.filter((o) => !hiddenKeys.includes(getRepositoryOptionHash(o)));
     }
-  }, [allRepos, hiddenKeys]);
+    const hiddenKeys = new Set(JSON.parse(hiddenKeysJSON));
+    return allRepos.filter((o) => !hiddenKeys.has(getRepositoryOptionHash(o)));
+  }, [allRepos, hiddenKeysJSON]);
```

Comment:
This fixes a bug I noticed where `hiddenKeys` is being updated in a loop with the same values. It appears to be happening because of the `onStateChanged`, which pushes updates to the hidden and visible keys, which results in an update to context, which pushes a new state to the filter component, which triggers `onStateChanged`, etc.

I think there's always a bit of risk with setting the global hidden/visible key state, as any caller might set the hidden repo keys at any time and at any cadence, and there's nothing stopping the entire workspace context from updating at that cadence.

We can at least guard against this flavor of loop (where the same list is set in state repeatedly) by stringifying the array and using that for `useMemo` comparison. This way, if the list of hidden and visible keys is the same, we won't update the `visibleRepos` array at all, and the filter component won't push a new state value.